perun-oidc-server/src/main/java/cz/muni/ics/oauth2/web/DeviceEndpoint.java

```diff
@@ -164,138 +210,230 @@ public String requestDeviceCode(@RequestParam("client_id") String clientId, @Req
-		// sort scopes for display based on the inherent order of system scopes
-		for (SystemScope s : systemScopes) {
-			if (scopes.contains(s)) {
-				sortedScopes.add(s);
-			}
-		}
+		model.put(CLIENT, client);
+		model.put(DC, dc);
 
-		// add in any scopes that aren't system scopes to the end of the list
-		sortedScopes.addAll(Sets.difference(scopes, systemScopes));
-
-		model.put("scopes", sortedScopes);
+		Set<SystemScope> sortedScopes = getSortedScopes(dc);
+		model.put(SCOPES, sortedScopes);
 
 		AuthorizationRequest authorizationRequest = oAuth2RequestFactory.createAuthorizationRequest(dc.getRequestParameters());
 
-		session.setAttribute("authorizationRequest", authorizationRequest);
-		session.setAttribute("deviceCode", dc);
+		session.setAttribute(AUTHORIZATION_REQUEST, authorizationRequest);
+		session.setAttribute(DEVICE_CODE_SESSION_ATTRIBUTE, dc);
 
-		return "approveDevice";
+		return tryToSetThemedApproveDevice(model, p, req, (DeviceCode) model.get(DC));
 	}
 
 	@PreAuthorize("hasRole('ROLE_USER')")
 	@RequestMapping(value = "/" + USER_URL + "/approve", method = RequestMethod.POST)
-	public String approveDevice(@RequestParam("user_code") String userCode,
-								@RequestParam(value = "user_oauth_approval") Boolean approve,
-								ModelMap model, Authentication auth, HttpSession session) {
-		AuthorizationRequest authorizationRequest = (AuthorizationRequest) session.getAttribute("authorizationRequest");
-		DeviceCode dc = (DeviceCode) session.getAttribute("deviceCode");
+	public String approveDevice(@RequestParam(USER_CODE) String userCode,
+								@RequestParam(value = USER_OAUTH_APPROVAL) Boolean approve,
+								Principal p,
+								HttpServletRequest req,
+								ModelMap model,
+								Authentication auth,
+								HttpSession session) {
+
+		AuthorizationRequest authorizationRequest = (AuthorizationRequest) session.getAttribute(AUTHORIZATION_REQUEST);
+		DeviceCode dc = (DeviceCode) session.getAttribute(DEVICE_CODE_SESSION_ATTRIBUTE);
 
 		// make sure the form that was submitted is the one that we were expecting
 		if (!dc.getUserCode().equals(userCode)) {
-			model.addAttribute("error", "userCodeMismatch");
-			return "requestUserCode";
+			model.addAttribute(ERROR, USER_CODE_MISMATCH);
+			return REQUEST_USER_CODE;
 		}
 
 		// make sure the code hasn't expired yet
 		if (dc.getExpiration() != null && dc.getExpiration().before(new Date())) {
-			model.addAttribute("error", "expiredUserCode");
-			return "requestUserCode";
+			model.addAttribute(ERROR, EXPIRED_USER_CODE);
+			return REQUEST_USER_CODE;
 		}
 
 		ClientDetailsEntity client = clientService.loadClientByClientId(dc.getClientId());
-
-		model.put("client", client);
+		model.put(CLIENT, client);
 
 		// user did not approve
 		if (!approve) {
-			model.addAttribute("approved", false);
-			return "deviceApproved";
+			model.addAttribute(APPROVED, false);
+			return tryToSetThemedApproveDevice(model, p, req, (DeviceCode) session.getAttribute(DEVICE_CODE));
```

Comment:
Compare the last parameter with the call made on line `279`. I believe you are fetching the same object, with different keys. Perhaps that should be unified.